doc/api/packages.md

```diff
@@ -83,6 +84,23 @@ future-proof the package in case the default type of Node.js ever changes, and
 it will also make things easier for build tools and loaders to determine how the
 files in the package should be interpreted.
 
+Node.js will refuse to load the following when passed to `node` as the
+initial input and the nearest parent `package.json` file contains a top-level
+[`"type"`][] field with a value of `"module"`:
+
+* Files whose name doesn't end in `.js`, `.mjs`, `.cjs`, or `.wasm`.
```

Comment:
Rather than saying what Node _won’t_ do, why not say what it will? I think it’s easier to understand the positive case than the negative:

```suggestion
When the nearest parent `package.json` file contains a top-level
[`"type"`][] field with a value of `"module"`, the `node` command will
accept as input only files with `.js`, `.mjs`, or `.cjs` extensions;
and with `.wasm` extensions when `--experimental-wasm-modules` is enabled.
```

(And please add a link to `--experimental-wasm-modules`; and is a wasm entry point actually supported?)

Reply:
The type field has no bearing on what node accepts - it accepts a .js file regardless, and has no effect on the others.

Reply:
That isn't true, type affects if the ESM module loader is used for the entry point: https://github.com/nodejs/node/blob/2cc7a91a5d855b4ff78f21f1bb8d4e55131d0615/lib/internal/modules/run_main.js#L41 . Also it makes extension-less files fail to load and that should probably be called out.

Reply:
For the latter, that’s good to know, and should indeed be called out.

For the former, are you saying `node foo` won’t do any extension lookup on `foo.cjs`,`foo.js`, or `foo.mjs`? If so, i understand why - but that’s not enabling ESM, that’s disabling the standard/default CJS loader, and that should definitely be explicitly called out as well (as a potential hazard/impact of using type module)

Reply:
under `type:"module"` the regular ESM semantics are applied against the preresolved filepath `argv[1]` absolute `file:` URL; so, no searching occurs. For `type:"commonjs"` the CJS semantics are applied against the preresolved `argv[1]` absolute file path.

Reply:
Thanks, TIL a second effect of type module :-) please do call this out explicitly; the more users correctly understand type module’s effects (which as i now understand it are just that .jjs is ESM + node CLI stops doing extension/index searching) the better.

Reply:
> under `type:"module"` the regular ESM semantics are applied against the preresolved filepath `argv[1]` absolute `file:` URL; so, no searching occurs.

I beg to differ:

```sh
mkdir repro
cd repro
echo 'console.log(this === undefined ? "ESM" : "CJS")' > file.js
node file # CJS
echo '{"type":"module"}' > package.json
node file # ESM
```

Reply:
ooo, this one is fun. it apparently got carved out some weird edge case that I missed but it doesn't do *all* kinds of searching? https://github.com/nodejs/node/blob/master/lib/internal/modules/esm/resolve.js#L1000 doesn't seem to be the path it is taking but it is hitting shouldUseESMLoader. Is this a bug?

Reply:
Seems more like a feature than a bug :-) removing it will likely be a breaking change, as users of node 12, 14, 16, and 17 are likely relying on it.

Reply:
so it looks like CJS resolver is always hit prior to the the ESM resolver, but things are weird and spaghetti https://github.com/nodejs/node/blob/e46c680bf2b211bbd52cf959ca17ee98c7f657f5/lib/internal/modules/run_main.js#L70 , so it does both, not one or the other when it thinks it is ESM. there is an odd comment about back compat, but I have no memory of a compat issue.

Reply:
@ljharb the comments in the file that are causing it explicitly state that removing extension searching can be done in the future.

Reply:
also, this seems like it would break if someone mucked with require.extensions in weird ways and passed it to ESM

Reply:
The comments were probably from when it was still experimental. Now that it no longer is, it doesn't seem like something that can be removed without a semver-major.

Reply:
Likely would be semver-major yes.